src/viam/sdk/resource/resource.hpp

```diff
@@ -51,5 +42,21 @@ class Resource {
     std::string name_;
 };
 
+class Stoppable {
```

Comment:
Good ideas for all the bulleted points; should be done.

> Meanwhile, I wonder a bit whether we intend to have things that are Stoppable that are not Resources. As written that is possible, and Stoppable functions basically as a mixin. Is it ever important given a Stoppable pointer to be able to treat it as a Resource?

I don't think anything besides a resource _should_ be `Stoppable`. With the current design, it seems like we do not need to treat a `Stoppable` pointer as a `Resource`, though, only the other way around.

> If so, you would probably want to have Stoppable derive virtually from Resource, but then you would also need to do that for Component and Service so you could still mix in Stoppable without ending up with two copies of the Resource base. That may be more trouble than it is worth though since Resource has state and a constructor that takes an argument.

So I initially tried something like this, the hierarchy was really odd. I found it hard to introduce a single new `StoppableResource` class derived from `Resource`; I found myself adding other classes like `StoppableComponent` and `StoppableService` to avoid two copies of the same base as you mention. If one of you can think of an uncomplicated hierarchy where `Stoppable` is a `Resource`, I'm all for it. I think the inability to treat a `Stoppable` as a `Resource` is not _terrible_ though.

Reply:
Yeah, I think it becomes more complex than necessary. Let's leave it as you have it: as a mixin class not related to `Resource`.